File: backend/src/test/java/ca/metricalsky/winston/test/faker/providers/entity/AuthorEntityProvider.java
package ca.metricalsky.winston.test.faker.providers.entity;

import ca.metricalsky.winston.entity.view.AuthorDetailsView;
import ca.metricalsky.winston.test.faker.WinstonFaker;
import lombok.Builder;
import lombok.Value;
import net.datafaker.providers.base.AbstractProvider;

public class AuthorEntityProvider
        extends AbstractProvider<WinstonFaker> {

    public AuthorEntityProvider(WinstonFaker faker) {
        super(faker);
    }

    public AuthorDetailsView authorDetails() {
        var totalCommentCount = faker.number().numberBetween(1L, 1000);

        return MockAuthorDetailsView.builder()
                .authorId(faker.youtube().authorId())
                .channelCount(faker.number().numberBetween(1L, 20))
                .videoCount(faker.number().numberBetween(1L, 100))
                .totalCommentCount(totalCommentCount)
                .replyCount(faker.number().numberBetween(0L, totalCommentCount - 1))
                .build();
    }

    @Value
    @Builder
    private static class MockAuthorDetailsView implements AuthorDetailsView {

        String authorId;
        Long channelCount;
        Long videoCount;
        Long totalCommentCount;
        Long replyCount;
    }
}

File: database/src/testFixtures/kotlin/ca/metricalsky/winston/database/test/faker/providers/AuthorEntityProvider.kt
package ca.metricalsky.winston.database.test.faker.providers

import ca.metricalsky.winston.database.entity.author.AuthorEntity
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import net.datafaker.providers.base.AbstractProvider

class AuthorEntityProvider(faker: DatabaseFaker) : AbstractProvider<DatabaseFaker>(faker) {

    fun id(): String = faker.regexify("UC[A-Za-z0-9_-]{22}")

    fun minimalEntity() = AuthorEntity(
        id = id(),
    )

    fun completeEntity() = AuthorEntity(
        id = id(),
        displayName = "@" + faker.name().firstName(),
        channelUrl = faker.internet().url(),
        profileImageUrl = faker.internet().url(),
        aliases = setOf("@" + faker.name().firstName()),
    )
}
